### PR TITLE
Test fix for network packet drops

### PR DESCRIPTION
**Pull request by Atlante45**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: d9d0c361997c61b3b414299ceeaaf5c080f4fba2 available at: https://gist.github.com/14090999e13e44008a07<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.dmg

File: libraries/networking/src/udt/SaltShaker.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+//
+
+#ifndef hifi_SaltShaker_h
+#define hifi_SaltShaker_h
+
+#include "Packet.h"
+
+namespace udt {
+
+class SaltShaker {
```
Comment (huffman) [inline libraries/networking/src/udt/SaltShaker.h]: I thought this was going to be named something different?

File: libraries/networking/src/udt/SaltShaker.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+using Key = uint64_t;
+static const std::array<Key, 4> KEYS {{
+    0x0,
+    0xd6ea42f07644016a,
+    0x700f7e3414dc4d8c,
+    0x54c92e8d2c871642
+}};
+
+void saltingHelper(char* start, int size, Key key) {
```
Comment (huffman) [inline libraries/networking/src/udt/SaltShaker.cpp]: More of a style thing, but I prefer functions to be verbs (getters that don't include `get` are sort of OK, they don't bug me as much). Someone else should probably weigh in. I think @birarda does this too, so I'm probably in the minority.

File: libraries/networking/src/udt/SaltShaker.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+public:
+    std::unique_ptr<Packet> salt(const Packet& packet, unsigned int saltiness);
+    void unsalt(Packet& packet, unsigned int saltiness);
```
Comment (huffman) [inline libraries/networking/src/udt/SaltShaker.h]: Should these be static? Seems weird that this object is being instantiated when used when it has 0 state.

File: libraries/networking/src/udt/SaltShaker.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    const auto end = start + size;
+
+    auto p = start;
+    for (; p < end; p += sizeof(Key)) {
+        *reinterpret_cast<Key*>(p) ^= key;
```
Comment (huffman) [inline libraries/networking/src/udt/SaltShaker.cpp]: What if size isn't a multiple of `sizeof(Key)` ? Seems like you could xor past the end of the data.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8a62aa5f6dc6fee4ce3edde9f60df1a7db4a8140 available at: https://gist.github.com/14090999e13e44008a07<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.dmg

File: libraries/networking/src/udt/SaltShaker.h (hunk 1 above)
Comment (Atlante45) [inline libraries/networking/src/udt/SaltShaker.h]: Yes, I just left it for for the tests. this is not PR ready yet.

File: libraries/networking/src/udt/Packet.cpp [hunk 5]
```diff
@@ -119,19 +155,12 @@ void Packet::writeMessageNumber(MessageNumber messageNumber, PacketPosition posi
     writeHeader();
 }
 
-void Packet::writeSequenceNumber(SequenceNumber sequenceNumber) const {
+void Packet::writeSequenceNumber(SequenceNumber sequenceNumber, ObfuscationLevel level) const {
```
Comment (huffman) [inline libraries/networking/src/udt/Packet.cpp]: Any reason this wasn't in a separate setter? Seems like that's what we do with other header values.

File: libraries/networking/src/udt/SaltShaker.cpp (hunk 4 above)
Comment (Atlante45) [inline libraries/networking/src/udt/SaltShaker.cpp]: No that's the point, this stops before the end.
And the next loop takes care of the rest.

File: libraries/networking/src/udt/SendQueue.cpp [hunk 6]
```diff
@@ -354,14 +359,44 @@ bool SendQueue::maybeResendPacket() {
...
+
+                    if (resendPacket.isPartOfMessage()) {
+                        debugString += "\n";
+                        debugString += "    Message Number: %1, Part Number: %2.";
+                        debugString = debugString.arg(QString::number(resendPacket.getMessageNumber()),
+                                                      QString::number(resendPacket.getMessagePartNumber()));
+                    }
+
+
+                    static QString repeatedMessage = LogHandler::getInstance().addRepeatedMessageRegex("^Obfuscating packet .{0,1000}");
```
Comment (birarda) [inline libraries/networking/src/udt/SendQueue.cpp]: This seems like an overly eager regex - looks for any single character between 0 or a 1000 times at the end?

File: libraries/networking/src/udt/SaltShaker.h (hunk 3 above)
Comment (Atlante45) [inline libraries/networking/src/udt/SaltShaker.h]: Might event get moved to BasePacket or Packet.

File: libraries/networking/src/udt/SendQueue.cpp (hunk 6 above)
Comment (birarda) [inline libraries/networking/src/udt/SendQueue.cpp]: I would be specific like `^Obfuscating packet \d+ with level \d+`

File: libraries/networking/src/udt/Packet.cpp (hunk 5 above)
Comment (Atlante45) [inline libraries/networking/src/udt/Packet.cpp]: No, we write on a per line basis. I could move it though.

File: libraries/networking/src/udt/SendQueue.cpp (hunk 6 above)
Comment (Atlante45) [inline libraries/networking/src/udt/SendQueue.cpp]: Yes, I had to do that, because Qt doesn't have anything that acts like ".*"

Comment (birarda) [inline libraries/networking/src/udt/SendQueue.cpp]: I see you have to deal with the optional message number stuff as well - I guess it's fine enough to use `.+`

File: libraries/networking/src/udt/SaltShaker.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    p -= sizeof(Key);
+
+    for (int i = 0; p < end; ++p || ++i) {
+        *p ^= *(reinterpret_cast<const char*>(&key) + i);
```
Comment (huffman) [inline libraries/networking/src/udt/SaltShaker.cpp]: I see, you xor the tail bits here. It still looks like you can xor past the end in the code above, but I could be wrong.

Comment (Atlante45) [inline libraries/networking/src/udt/SaltShaker.cpp]: Oh yeah, I think you're right.

**Comment by Atlante45** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6a2a76ee885739bfffefd256474616b551770bf9 available at: https://gist.github.com/14090999e13e44008a07<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6a2a76ee885739bfffefd256474616b551770bf9 available at: https://gist.github.com/14090999e13e44008a07<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7146.dmg

**Comment by Atlante45** [comment]
Good news, that works! I'm gonna clean up the code and make a new PR.

**Pull request closed without merging**